microsetta_public_api/models/_alpha.py

```diff
@@ -0,0 +1,221 @@
+    may represent the alpha values of many samples (such as all fecal samples).
+
+    Attributes
+    ----------
+    name : str
+        The name of the group.
+    metric : str
+        The name of the metric expressed.
+    data : dict
+        The sample to value data.
+    """
+    __slots__ = ()
+
+    def __init__(self, *args, **kwargs):
+        if args:
+            raise NotImplementedError("%s only supports kwargs" %
+                                      str(self.__class__))
+        super()
+
+    def to_dict(self) -> Dict:
+        return self._asdict()
+
+    def __str__(self) -> str:
+        return str(self.to_dict())
+
+
+class GroupAlpha(_ga_named):
+    """Minimal information to characterize alpha detail of a group
+
+    A group may represent alpha detail of a single sample. Or, a group
+    may represent the summarized alpha distribution of many samples (such
+    as all fecal samples).
+
+    Attributes
+    ----------
+    name : str
+        The name of the group (e.g., 'sample-foo').
+    metric : str
+        The name of the metric expressed.
+    mean : float
+        The mean of the group values.
+    median : float
+        The median of the group values.
+    std : float
+        The std of the group values.
+    group_size : int
+        The number of samples represented.
+    percentile : list of int, optional
+        The percentiles represented (e.g., 25th, 50th, 75th, etc) or None if
+        the group_size is 1.
+    percentile_values : list of float, optional
+        The values of the percentiles or None if the group_size is 1.
+    """
+    __slots__ = ()
+
+    def __init__(self, *args, **kwargs):
+        if args:
+            raise NotImplementedError("%s only supports kwargs" %
+                                      str(self.__class__))
+
+        if kwargs['group_size'] == 1:
+            if not np.allclose(kwargs['mean'], kwargs['median']):
+                raise ValueError('Data non-sensical for n=1.')
+
+            if not np.allclose(kwargs['std'], 0.0):
+                raise ValueError('Data non-sensical for n=1.')
+
+            if kwargs['percentile'] is not None:
+                raise ValueError('Data non-sensical for n=1.')
+
+            if kwargs['percentile_values'] is not None:
+                raise ValueError('Data non-sensical for n=1.')
+
+        elif kwargs['group_size'] > 1:
+            if kwargs['percentile'] is None:
+                raise ValueError('Missing percentiles.')
+
+            if kwargs['percentile_values'] is None:
+                raise ValueError('Missing percentiles.')
+
+            if len(kwargs['percentile']) != len(kwargs['percentile_values']):
+                raise ValueError('Passing in unmatched percentiles.')
+
+        else:
+            raise ValueError('Instantiation with a bad group_size.')
+
+        super()
+
+    def to_dict(self) -> Dict:
+        return self._asdict()
+
+    def __str__(self) -> str:
+        return str(self.to_dict())
```

Comment:
Is this function here to help with json serialization, or just to print for debugging?  Our to_api approach for json serialization was helpful to us because it gave us a place to transform our model classes to match the api.  But it suffered when attempting to nest objects, as the json serialization calls didn't know to recursively call the method.  For building strings this should be totally fine though

Reply:
This one is just for building strings for now.

It seems like what you are saying is to be careful of converting objects to dicts if they may contain references to other objects? When interfacing with flask.

Reply:
It probably would be good to have this go through `json.dumps` rather than `str` but I think that can be deferred until these objects are hooked up. Note that [`_asdict`](https://docs.python.org/3/library/collections.html#collections.somenamedtuple._asdict) is a member method of `namedtuple`, is built in and despite the leading `_` it is not intended to be private. Assuming developers/users are abiding by the docstrings, none of the types used should be problematic for serialization.